pyccel/ast/operators.py

```diff
@@ -1395,27 +1351,44 @@ def _calculate_type(cond, value_true, value_false):
         DataType
             The Python type of the object.
         """
-        if value_true.dtype is value_false.dtype and value_true.class_type is value_false.class_type:
+        if value_true.class_type is value_false.class_type:
             return value_true.class_type
 
         try:
             class_type = value_true.class_type + value_false.class_type
         except NotImplementedError:
             raise TypeError(f'Cannot determine the type of ({value_true}, {value_false})') #pylint: disable=raise-missing-from
 
+        if value_false.class_type.order != value_true.class_type.order :
+            errors.report('Ternary Operator results should have the same order', severity='fatal')
+
         return class_type
 
     @staticmethod
-    def _calculate_shape_rank(cond, value_true, value_false):
+    def _calculate_shape(cond, value_true, value_false):
         """
-        Sets the shape and rank and the order for IfTernaryOperator
+        Calculate the shape of the result.
+
+        Calculate the shape of the result of the IfTernaryOperator.
+        The shape is equal to the shape of one of the outputs.
+
+        Parameters
+        ----------
+        cond : TypedAstNode
+            The first argument passed to the operator representing the condition.
+        value_true : TypedAstNode
+            The second argument passed to the operator representing the result if the
+            condition is true.
+        value_false : TypedAstNode
+            The third argument passed to the operator representing the result if the
+            condition is false.
+
+        Returns
+        -------
+        tuple[TypedAstNode]
+            The shape of the resulting object.
         """
-        shape = value_true.shape
-        rank  = value_true.rank
-        if rank is not None and rank > 1:
-            if value_false.order != value_true.order :
-                errors.report('Ternary Operator results should have the same order', severity='fatal')
-        return shape, rank
+        return value_true.shape
```

Comment:
Generally speaking, it is rarely possible to check if the shape of the arguments is compatible. That said there is a check in the constructor:
https://github.com/pyccel/pyccel/blob/13906227d05489514fa68bc4c34b6697e04176c3/pyccel/ast/operators.py#L1370-L1371

It could be moved here but I don't know if there is any reason to?

Reply:
I had not noticed! 😅